### PR TITLE
INT 1957 Cart Tax Calculation Validator

### PR DESCRIPTION
**Pull request by dallendalton**
This PR contains the logic for validating that our plugin should and can perform a tax calculation API request. It also includes test coverage for this functionality.

This PR will be merged into the feature branch which will later be released when all the functionality around cart tax calculation has been refactored.

### CODE REVIEW

**Review by jordan-k-johnson — APPROVED** [review]
LGTM, just left one minor comment!

File: tests/specs/tax-calculation/test-cart-tax-calculation-validator.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		$cart_tax_calculation_validator = new Cart_Tax_Calculation_Validator( $cart, $this->nexus_stub );
+		add_filter( 'taxjar_should_calculate_cart_tax', function( $order ) {
+			return false;
+		});
+
+		$this->expectException( Tax_Calculation_Exception::class );
+		$cart_tax_calculation_validator->validate( $this->tax_request_body_stub );
+	}
+
+	public function test_no_nexus_throws_exemption() {
```
[the remark this answers is not in the document]
Comment (sethobey) [reply tests/specs/tax-calculation/test-cart-tax-calculation-validator.php:54]: +1

File: tests/specs/tax-calculation/test-cart-tax-calculation-validator.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	public function test_vat_exempt_cart_throws_exception() {
+		$cart = Cart_Builder::a_cart()->with_vat_exemption()->build();
+		$this->nexus_stub->method( 'has_nexus_check' )->willReturn( true );
+		$cart_tax_calculation_validator = new Cart_Tax_Calculation_Validator( $cart, $this->nexus_stub );
+
+		$this->expectException( Tax_Calculation_Exception::class );
+		$cart_tax_calculation_validator->validate( $this->tax_request_body_stub );
+	}
+
+	public function test_zero_cart_total_throws_exemption() {
```
Comment (sethobey) [inline tests/specs/tax-calculation/test-cart-tax-calculation-validator.php:72]: Also here, as Jordan mentioned above. "throws exception"

File: tests/specs/tax-calculation/test-order-tax-calculation-validator.php [hunk 3]
```diff
@@ -20,7 +19,7 @@ public function setUp() {
 		$this->mock_order = $this->createMock( WC_Order::class );
 		$this->mock_order->method( 'get_subtotal' )->willReturn( 1.0 );
 		$this->mock_tax_request_body = $this->createMock( Tax_Request_Body::class );
-		$this->mock_nexus            = $this->createMock( WC_Taxjar_Nexus::class );
+		$this->mock_nexus            = $this->createMock( \WC_Taxjar_Nexus::class );
```
Comment (sethobey) [inline tests/specs/tax-calculation/test-order-tax-calculation-validator.php:22]: Just wondered why you removed the "use" import here, if it was on purpose.

**Review by sethobey — CHANGES_REQUESTED** [review]

File: tests/specs/tax-calculation/test-cart-tax-calculation-validator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dallendalton) [reply tests/specs/tax-calculation/test-cart-tax-calculation-validator.php:54]: Good catch, I'll get these fixed.

File: tests/specs/tax-calculation/test-order-tax-calculation-validator.php (hunk 3 above)
Comment (dallendalton) [reply tests/specs/tax-calculation/test-order-tax-calculation-validator.php:22]: Nope it was not intentional. I will revert it.

**Review by sethobey — APPROVED** [review]

**Pull request merged**